frame/atomic-swap/src/lib.rs

```diff
@@ -0,0 +1,280 @@
+		SwapCancelled(AccountId, HashedProof),
+	}
+);
+
+decl_module! {
+	/// Module definition of atomic swap pallet.
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		type Error = Error<T>;
+
+		fn deposit_event() = default;
+
+		/// Register a new atomic swap, declaring an intention to send funds from origin to target
+		/// on the current blockchain. The target can claim the fund using the revealed proof. If
+		/// the fund is not claimed after `duration` blocks, then the sender can cancel the swap.
+		///
+		/// The dispatch origin for this call must be _Signed_.
+		///
+		/// - `target`: Receiver of the atomic swap.
+		/// - `hashed_proof`: The blake2_256 hash of the secret proof.
+		/// - `balance`: Funds to be sent from origin.
+		/// - `duration`: Locked duration of the atomic swap.
+		#[weight = T::DbWeight::get().reads_writes(1, 1).saturating_add(40_000_000)]
+		fn create_swap(
+			origin,
+			target: AccountIdFor<T>,
+			hashed_proof: HashedProof,
+			balance: BalanceFor<T>,
+			duration: BlockNumberFor<T>,
+		) {
+			let source = ensure_signed(origin)?;
+			ensure!(
+				!PendingSwaps::<T>::contains_key(&target, hashed_proof),
+				Error::<T>::AlreadyExist
+			);
+
+			T::Currency::reserve(&source, balance)?;
+
+			let swap = PendingSwap {
+				source,
+				balance,
+				end_block: frame_system::Module::<T>::block_number() + duration,
+				claimed: false,
+			};
+			PendingSwaps::<T>::insert(target.clone(), hashed_proof.clone(), swap.clone());
+
+			Self::deposit_event(
+				RawEvent::NewSwap(target, hashed_proof, swap)
+			);
+		}
+
+		/// Claim an atomic swap.
+		///
+		/// The dispatch origin for this call must be _Signed_.
+		///
+		/// - `proof`: Revealed proof of the claim.
+		#[weight = T::DbWeight::get().reads_writes(2, 2)
+		  .saturating_add(40_000_000)
+		  .saturating_add((proof.len() as Weight).saturating_mul(100))
+		]
+		fn claim_swap(
+			origin,
+			proof: Vec<u8>,
+		) -> DispatchResult {
+			ensure!(
+				proof.len() <= T::ProofLimit::get() as usize,
+				Error::<T>::ProofTooLarge,
+			);
+
+			let target = ensure_signed(origin)?;
+			let hashed_proof = blake2_256(&proof);
+
+			let mut swap = PendingSwaps::<T>::get(&target, hashed_proof)
+				.ok_or(Error::<T>::InvalidProof)?;
+			swap.claimed = true;
+
+			match T::Currency::repatriate_reserved(
+				&swap.source,
+				&target,
+				swap.balance,
+				BalanceStatus::Free,
+			) {
+				Err(e) => {
+					let expired = frame_system::Module::<T>::block_number() >
+						swap.end_block.saturating_add(T::ExpireDuration::get());
+					if expired {
+						PendingSwaps::<T>::remove(target.clone(), hashed_proof.clone());
+
+						Self::deposit_event(
+							RawEvent::SwapClaimFailed(target, hashed_proof, false)
+						);
+					} else {
+						PendingSwaps::<T>::insert(target.clone(), hashed_proof.clone(), swap);
+
+						Self::deposit_event(
+							RawEvent::SwapClaimFailed(target, hashed_proof, true)
+						);
+					}
+
+					Err(e.into())
```

Comment:
You shouldnt return error and change the state of the blockchain. Why did you subtly change the design of your blockchain this way?

Every time your pallets returns Err it should be a noop, and this should be verified with runtime tests.